sql/parser/builtins.go

```diff
@@ -436,7 +436,11 @@ var builtins = map[string][]builtin{
 			types:      typeList{},
 			returnType: DummyDate,
 			fn: func(e EvalContext, args DTuple) (Datum, error) {
-				return MakeDDate(e.StmtTimestamp.Time), nil
+				loc, err := e.GetLocation()
+				if err != nil {
+					return DummyTimestamp, err
+				}
+				return MakeDDate(e.StmtTimestamp.Time.In(loc)), nil
```

Comment:
I still think making this lazy is unnecessarily error-prone. If we set the right time zone in `StmtTimestamp` then we don't have to evaluate all potential uses of the variable to see whether they need to apply the time zone or not. It's simple to add a cache around `time.LoadLocation` when and if we need to.